{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "starling_ped = \"./starling-data/final_filtering/starling_noduprel_qual_miss_filt-eop-mod__NO_HEADER.plink.ped\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "starling_df = pd.read_csv(starling_ped, sep='\\t', dtype=str, header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "indiv_1_data = list(starling_df.iloc[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "from enum import IntEnum\n",
    "\n",
    "class AlleleBase(IntEnum):\n",
    "    A = 0\n",
    "    T = 1\n",
    "    G = 3\n",
    "    C = 4\n",
    "    N = 5 # missing allele\n",
    "\n",
    "    def __str__(self):\n",
    "        return self.name\n",
    "\n",
    "class string(str):\n",
    "    def toAllele(self) -> object:\n",
    "        # convert string astr to AlleleBase\n",
    "        self = self.strip().upper()\n",
    "        if self == 'A':\n",
    "            return AlleleBase.A\n",
    "        if self == 'T':\n",
    "            return AlleleBase.T\n",
    "        if self == 'G':\n",
    "            return AlleleBase.G\n",
    "        if self == 'C':\n",
    "            return AlleleBase.C\n",
    "        if self == 'N' or self == '0' or self.upper() == \"NAN\":\n",
    "            return AlleleBase.N\n",
    "        raise Exception(f\"Allele must be A, T, G, C, or N, received {self}\")\n",
    "\n",
    "\n",
    "class SNP():\n",
    "    '''\n",
    "    Will return True for missing SNPs like NN-NN or NT-NA\n",
    "    '''\n",
    "    def __init__(self, a1: AlleleBase, a2: AlleleBase):\n",
    "        self._a1 = a1\n",
    "        self._a2 = a2\n",
    "        if a1 > a2:\n",
    "            self._a1 = a2 \n",
    "            self._a2 = a1\n",
    "\n",
    "    def __eq__(self, other: AlleleBase) -> bool:\n",
    "        if self._a1 == other._a1 and self._a2 == other._a2:\n",
    "            return True\n",
    "        if self._a1 == other._a2 and self._a2 == other._a1:\n",
    "            # expect a1, a2, to be ordered by ctor, but check anyway, \n",
    "            # could have unexpected override\n",
    "            return True\n",
    "        return False\n",
    "\n",
    "    # in Python 3, if __ne__ is undefined, it defaults to inverse of __eq__\n",
    "\n",
    "    def __str__(self) -> str:\n",
    "        return f\"{str(self._a1)}{str(self._a2)}\"\n",
    "\n",
    "    def __len__(self):\n",
    "        # number of alleles\n",
    "        return 2\n",
    "\n",
    "    def __getitem__(self, idx):\n",
    "        if idx == 0:\n",
    "            return self._a1\n",
    "        if idx == 1 or idx == -1:\n",
    "            return self._a2\n",
    "        raise Exception(\"index out of bounds\")\n",
    "\n",
    "class StrictSNP(SNP):\n",
    "    '''\n",
    "    a SNP, but equals method excludes missing SNPs\n",
    "    False for missing SNPs like NN-NN or NT-NA\n",
    "    '''\n",
    "    def __init__(self, snp: SNP):\n",
    "        self._a1 = snp._a1\n",
    "        self._a2 = snp._a2\n",
    "        if self._a1 > self._a2:\n",
    "            tmp = self._a1\n",
    "            self._a1 = self._a2 \n",
    "            self._a2 = tmp \n",
    "\n",
    "    def __eq__(self, other: AlleleBase) -> bool:\n",
    "        AB = AlleleBase\n",
    "        if self._a1 == AB.N or other._a1 == AB.N:\n",
    "            # since a1, a2 are ordered, if a1 is N then they both must be\n",
    "            return False\n",
    "        if self._a2 == AB.N and other._a2 == AB.N:\n",
    "            # both are haplotypes, so check equality\n",
    "            return self._a1 == other._a1\n",
    "        # other cases:\n",
    "        #   - one is a haplotype and one has both SNPs\n",
    "        #   - no unknown SNPs present\n",
    "        return SNP(self._a1, self._a2) == SNP(other._a1, other._a2)\n",
    "\n",
    "\n",
    "class Individual():\n",
    "    '''\n",
    "    for use explicitly with PED files\n",
    "    '''\n",
    "    def __init__(self, data: list):\n",
    "        # this assumes a PED file\n",
    "        self.family_id = data[0]\n",
    "        self.individual_id = data[1]\n",
    "        # list of SNPs\n",
    "        self._SNPs = self.process_snps(data[6:])\n",
    "\n",
    "    def process_snps(self, data):\n",
    "        # convert list of nucleotides to SNPs\n",
    "        snps = [None] * int(len(data) / 2) # instantiate mutable list\n",
    "        sidx = 0\n",
    "        for idx in range(0, len(data), 2): # iterate ever 2 indices\n",
    "            a1 = string(data[idx]).toAllele()\n",
    "            a2 = string(data[idx+1]).toAllele()\n",
    "            snps[sidx] = SNP(a1, a2)\n",
    "            sidx += 1\n",
    "        return snps\n",
    "\n",
    "    def __str__(self):\n",
    "        return str([str(snp) for snp in self._SNPs])\n",
    "\n",
    "    def __eq__(self, other):\n",
    "        if self.family_id != other.family_id:\n",
    "            return False\n",
    "        if self.individual_id != other.individual_id:\n",
    "            return False\n",
    "\n",
    "        return str(self) == str(other)\n",
    "\n",
    "    def __len__(self):\n",
    "        return len(self._SNPs)\n",
    "\n",
    "    def __getitem__(self, idx):\n",
    "        return self._SNPs[idx]\n",
    "\n",
    "\n",
    "class BetaDiverstiy(Individual):\n",
    "    def __init__(self, indiv: Individual):\n",
    "        #super().__init__()\n",
    "        self.family_id = indiv.family_id\n",
    "        self.individual_id = indiv.individual_id\n",
    "        # list of SNPs\n",
    "        self._SNPs = indiv._SNPs\n",
    "\n",
    "    def jaccard(self, other: Individual):\n",
    "        if self == other:\n",
    "            return 0 # no differentiation\n",
    "\n",
    "        def noMatchFound(snp1: StrictSNP, snp2: StrictSNP) -> bool :\n",
    "            s11 = snp1[0]\n",
    "            s12 = snp1[1]\n",
    "            s21 = snp2[0]\n",
    "            s22 = snp2[1]\n",
    "            if s11 == s21 and s11 != AlleleBase.N:\n",
    "                return False\n",
    "            if s12 == s21 and s12 != AlleleBase.N:\n",
    "                return False\n",
    "            if s11 == s22 and s11 != AlleleBase.N:\n",
    "                return False\n",
    "            if s12 == s22 and s12 != AlleleBase.N:\n",
    "                return False\n",
    "            return True\n",
    "\n",
    "        a_joint_presences = 0\n",
    "        b_c_half_match = 0\n",
    "        d_joint_absence = 0\n",
    "        for idx in range(0, len(self)):\n",
    "            snp1 = StrictSNP(self[idx])\n",
    "            snp2 = StrictSNP(other[idx])\n",
    "            if snp1 == snp2:\n",
    "                a_joint_presences += 1\n",
    "            else:\n",
    "                if noMatchFound(snp1, snp2):\n",
    "                    d_joint_absence += 1\n",
    "                else:\n",
    "                    b_c_half_match += 1\n",
    "\n",
    "        numerator = a_joint_presences\n",
    "        denominator = a_joint_presences + b_c_half_match\n",
    "        result = 1 - numerator / denominator\n",
    "        return result\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(starling_df)\n",
    "indivs = [None] * len(starling_df) # create vector of individual objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "starling_np = starling_df.to_numpy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a list of Individual objects\n",
    "indiv_idx = 0\n",
    "for row_idx in range(0, len(starling_np)):\n",
    "    indiv = Individual(starling_np[row_idx])\n",
    "    indivs[indiv_idx] = indiv\n",
    "    indiv_idx += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# individual -> location mapping\n",
    "\n",
    "indiv2site = {}\n",
    "indiv2site[\"ATB\"] = \"ANT (BE)\"\n",
    "indiv2site[\"AUK\"] = \"AUK (NZ)\"\n",
    "indiv2site[\"HAM\"] = \"ANT (BE)\"\n",
    "indiv2site[\"BAD\"] = \"BLN (NZ)\"\n",
    "indiv2site[\"BRA\"] = \"BLN (NZ)\"\n",
    "indiv2site[\"CAN\"] = \"CAN (NZ)\"\n",
    "indiv2site[\"MONK\"] = \"MKW (UK)\"\n",
    "indiv2site[\"sv\"] = \"MLV (AU)\"\n",
    "indiv2site[\"ABGYY\"] = \"NWC (UK)\"\n",
    "indiv2site[\"BPBBv2\"] = \"NWC (UK)\"\n",
    "indiv2site[\"BRYY\"] = \"NWC (UK)\"\n",
    "indiv2site[\"BYPP\"] = \"NWC (UK)\"\n",
    "indiv2site[\"GYRR\"] = \"NWC (UK)\"\n",
    "indiv2site[\"PPYY\"] = \"NWC (UK)\"\n",
    "indiv2site[\"PRRR\"] = \"NWC (UK)\"\n",
    "indiv2site[\"PYYY\"] = \"NWC (UK)\"\n",
    "indiv2site[\"V10181\"] = \"NWC (UK)\"\n",
    "indiv2site[\"V10531\"] = \"NWC (UK)\"\n",
    "indiv2site[\"V10731A\"] = \"NWC (UK)\"\n",
    "indiv2site[\"NOR\"] = \"ORG (AU)\"\n",
    "indiv2site[\"PLM\"] = \"PLM (NZ)\"\n",
    "indiv2site[\"UHT\"] = \"UHT (NZ)\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_sites = [\n",
    "    \"ANT (BE)\", \"AUK (NZ)\", \"BLN (NZ)\", \n",
    "    \"CAN (NZ)\", \"MKW (UK)\", \"MLV (AU)\", \n",
    "    \"NWC (UK)\", \"ORG (AU)\", \"PLM (NZ)\", \"UHT (NZ)\"\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a nested dictionary with values for each pair of individuals \n",
    "# then loop through it later and summarize all pairs per pair of sites\n",
    "nested_results = {}\n",
    "compared = {}\n",
    "for i in range(0, len(starling_np)):\n",
    "    bd = BetaDiverstiy(indivs[i])\n",
    "    for j in range(0, len(starling_np)):\n",
    "        if i == j:\n",
    "            continue\n",
    "        indiv = indivs[j]\n",
    "        key_i = indiv2site[starling_np[i][0]] # site i\n",
    "        key_j = indiv2site[starling_np[j][0]] # site j\n",
    "        if key_i not in nested_results:\n",
    "            nested_results[key_i] = {}\n",
    "        if key_j not in nested_results[key_i]:\n",
    "            nested_results[key_i][key_j] = []\n",
    "        nested_results[key_i][key_j].append(bd.jaccard(indiv))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"heatmap5_median.csv\", \"w\") as f:\n",
    "    f.write(\"site1,site2,value\\n\")\n",
    "    for key_i in nested_results:\n",
    "        for key_j in nested_results:\n",
    "            if key_j not in nested_results[key_i]:\n",
    "                print(f\"missing: {key_i}-{key_j}\")\n",
    "            else:\n",
    "                f.write(f\"{key_i},{key_j},{np.median(nested_results[key_i][key_j])}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"heatmap5_mean.csv\", \"w\") as f:\n",
    "    f.write(\"site1,site2,value\\n\")\n",
    "    for key_i in nested_results:\n",
    "        for key_j in nested_results:\n",
    "            if key_j not in nested_results[key_i]:\n",
    "                print(f\"missing: {key_i}-{key_j}\")\n",
    "            else:\n",
    "                f.write(f\"{key_i},{key_j},{np.mean(nested_results[key_i][key_j])}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot in R:\n",
    "\n",
    "#library(ggplot2)\n",
    "#starling_df <- read.csv(\"heatmap5_mean.csv\")\n",
    "#\n",
    "#desired_order <- c(\"UHT (NZ)\", \"MLV (AU)\", \"PLM (NZ)\", \"ORG (AU)\", \"MKW (UK)\", \"CAN (NZ)\", \"BLN (NZ)\", \"AUK (NZ)\", \"ANT (BE)\", \"NWC (UK)\")\n",
    "##desired_order <- rev(desired_order)\n",
    "#starling_df$pop1 <- factor(starling_df$site1, levels = desired_order)\n",
    "#starling_df$pop2 <- factor(starling_df$site2, levels = desired_order)\n",
    "#\n",
    "#\n",
    "#\n",
    "#ggplot(starling_df, aes(x = pop1, y = pop2, fill = value, label = round(value, 3))) +\n",
    "#  geom_tile() +\n",
    "#  geom_text(color = \"black\", size = 3) +  \n",
    "#  scale_fill_gradient(low = \"white\", high = \"blue\") + \n",
    "#  theme_minimal() +\n",
    "#  theme(panel.grid = element_blank(), axis.title = element_blank()) +\n",
    "#  coord_fixed(ratio = 1, xlim = c(0.5, length(desired_order) + 0.5), ylim = c(0.5, length(desired_order) + 0.5)) +\n",
    "#  theme(axis.text.x = element_text(angle = 45, hjust = 1), axis.text.y = element_text(angle = 0, hjust = 1)) +\n",
    "#  geom_tile(data = subset(starling_df, pop1 == pop2), fill = \"white\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "slimutils2",
   "language": "python",
   "name": "venv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
