{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Starling Project Santure lab group, lead Kat Stuart ###\n",
    "\n",
    "### Kat's Project notes:\n",
    "### PSMC requires whole genome data of one individual per PSMC track. For this you will need:\n",
    "# Genome: /nesi/nobackup/uoa02613/kstuart_projects/At4_MynaStarling/data/resources/genomes/Svulgaris_vAU_1.0.fasta\n",
    "# Mapped WGS data: /nesi/nobackup/uoa02613/kstuart_projects/At4_MynaStarling/data/mapped_reads/starling_wgs_mapped/\n",
    "# Repeat bed file (for excluding repeat regions during var calling): /nesi/nobackup/uoa02613/kstuart_projects/Sv10_NZstarlings/data/repeat_analysis/Svulgaris_vAU_1.0.fasta_rm.bed\n",
    "\n",
    "# Essentially the way the code works is that for each sample you chose you call all the variants for that individually, and then run a separate PSMC analysis on each individual's SNP data. \n",
    "# I would aim to run a few different samples, maybe (looking in the 'Mapped WGS data' directory listed above) pick 3 x UK samples, 3 from NZ (S_BAD samples -- mix male and female) and 3x from one Australian sampling location (maybe auX_mai and auX_men samples).\n",
    "\n",
    "### Purpose of manuscript\n",
    "# Invasion history - different patterns in NZ vs Australia/UK\n",
    "# Demographic history, booms and busts and modelling to be done to test invasion history\n",
    "\n",
    "\n",
    "### Subsampled individuals from each pop; /nesi/nobackup/uoa02613/kstuart_projects/At4_MynaStarling/data/mapped_reads/starling_wgs_mapped/sub_sample\n",
    "### Au mai\n",
    "au03_mai\n",
    "au04_mai\n",
    "au27_mai\n",
    "### Au men\n",
    "au05_men\n",
    "au06_men\n",
    "au15_men\n",
    "### NZ - S_BAD\n",
    "S_BAD_001M\n",
    "S_BAD_007F\n",
    "S_BAD_008F\n",
    "### UK \n",
    "uk04_nwc\n",
    "uk06_nwc\n",
    "uk07_nwc\n",
    "\n",
    "#### STEP 1: Prepeare files for PSMC\n",
    "# 1) Call consensus variants per individual and index output. This will also mask sites overlapping repeat annotations.\n",
    "# 2) generate diploid sequences\n",
    "# 3) Convert to psmcfa file format\n",
    "\n",
    "#!/bin/bash -e\n",
    "#SBATCH --job-name=psmc_prep_05Mar2024\n",
    "#SBATCH --account=uoa02613\n",
    "#SBATCH --time=12:00:00\n",
    "#SBATCH --mem=10GB\n",
    "#SBATCH --output=MCN_%j.errout\n",
    "#SBATCH --mail-user=m.nehmens@massey.ac.nz\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=2\n",
    "#SBATCH --profile task\n",
    "#SBATCH --array=1-12\n",
    "\n",
    "SAMPLE=$(sed \"${SLURM_ARRAY_TASK_ID}q;d\" /nesi/nobackup/uoa02613/kstuart_projects/At4_MynaStarling/analysis/PSMC/psmc_indv_subset.txt)\n",
    "\n",
    "module load psmc/0.6.5-gimkl-2018b\n",
    "module load VCFtools/0.1.15-GCC-9.2.0-Perl-5.30.1\n",
    "module load SAMtools/1.19-GCC-12.3.0\n",
    "module load tabix/0.2.6-GCCcore-9.2.0\n",
    "module load BCFtools/1.19-GCC-11.3.0\n",
    "module load BBMap/39.01-GCC-11.3.0\n",
    "\n",
    "cd /nesi/nobackup/uoa02613/kstuart_projects/At4_MynaStarling/analysis/PSMC/fastq\n",
    "\n",
    "GENOME=/nesi/nobackup/uoa02613/kstuart_projects/At4_MynaStarling/data/resources/genomes/Svulgaris_vAU_1.0.fasta\n",
    "BAM_DIR=/nesi/nobackup/uoa02613/kstuart_projects/At4_MynaStarling/data/mapped_reads/starling_wgs_mapped/sub_sample\n",
    "REP_BED=/nesi/nobackup/uoa02613/kstuart_projects/Sv10_NZstarlings/data/repeat_analysis/Svulgaris_vAU_1.0.fasta_rm.bed\n",
    "\n",
    "#1\n",
    "bcftools mpileup -C 50 -q 20 -Q 25 -Ou -f ${GENOME} ${BAM_DIR}/${SAMPLE}.sorted.dup.bam | bcftools call -c -Oz -o ./${SAMPLE}.vcf.gz\n",
    "bcftools filter --SnpGap 10 -i \"DP>=5 & DP<=50\" ./${SAMPLE}.vcf.gz -Oz -o ./${SAMPLE}.filter.vcf.gz\n",
    "bcftools view --exclude-types indels ./${SAMPLE}.filter.vcf.gz -T ^${REP_BED} | bcftools sort --temp-dir ./tmp_${SAMPLE} -Oz -o ./${SAMPLE}.sort.vcf.gz\n",
    "tabix -p vcf ./${SAMPLE}.sort.vcf.gz\n",
    "\n",
    "#2\n",
    "bcftools view ./${SAMPLE}.vcf.gz | vcfutils.pl vcf2fq | gzip > ./${SAMPLE}.fastq.gz\n",
    "\n",
    "#3\n",
    "cd /nesi/nobackup/uoa02613/kstuart_projects/At4_MynaStarling/analysis/PSMC/\n",
    "fq2psmcfa -q 20 ./fastq/${SAMPLE}.fastq.gz > ./${SAMPLE}.psmcfa\n",
    "\n",
    "#### Submitted as job 44205319 on 05 March 2024 - produced jobs: 44205321 44205322 44205323 44205324 44205325 44205326 44205327 44205328 44205329 44205330 44205331, due to array\n",
    "# Job Wall-time:   76.3%  09:09:21 of 12:00:00 time limit\n",
    "# CPU Efficiency: 146.3%  13:23:48 of 09:09:21 core-walltime\n",
    "# Mem Efficiency:   7.1%  731.36 MB of 10.00 GB\n",
    "# Can reduce time to 10hrs, bumb CPU to 4, and reduce memory to 3GB\n",
    "\n",
    "### Step 2: Running PSMC from step 1 output\n",
    "# Time segment patterns:\n",
    "# 4+5*3+4 (default)\n",
    "# 4+10*3+6+8\n",
    "# 4+25*2+4+6\n",
    "# 4+30*2+4+6+10 ### time segment pattern here was chosen from Nadachowska-Brzyska_etal_2015_Supplemental_Table_S1, and paper's test for all bird species for which this works\n",
    "\n",
    "#!/bin/bash -e\n",
    "#SBATCH --job-name=psmc_main_06Mar2024\n",
    "#SBATCH --account=uoa02613\n",
    "#SBATCH --time=00-12:00:00\n",
    "#SBATCH --mem=2GB\n",
    "#SBATCH --output=MCN_%j.errout\n",
    "#SBATCH --mail-user=m.nehmens@massey.ac.nz\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=2\n",
    "#SBATCH --profile task\n",
    "#SBATCH --array=1-12\n",
    "\n",
    "SAMPLE=$(sed \"${SLURM_ARRAY_TASK_ID}q;d\" /nesi/nobackup/uoa02613/kstuart_projects/At4_MynaStarling/analysis/PSMC/psmc_indv_subset.txt)\n",
    "echo \"working with sample:\" $SAMPLE\n",
    "\n",
    "cd /nesi/nobackup/uoa02613/kstuart_projects/At4_MynaStarling/analysis/PSMC/\n",
    "\n",
    "module load psmc/0.6.5-gimkl-2018b\n",
    "\n",
    "psmc -N30 -t5 -r5 -p \"4+30*2+4+6+10\" -o ./results/${SAMPLE}.diploid.psmc ./${SAMPLE}.psmcfa\n",
    "\n",
    "#### Submitted as job 44220855 on 06 March 2024 and then jobs 44220856  44220857  44220858  44220859  44220860  44220861  44220862  44220863  44220864  44220865  44220866\n",
    "# Job Wall-time:   16.2%  01:56:50 of 12:00:00 time limit\n",
    "# CPU Efficiency: 100.0%  01:56:49 of 01:56:50 core-walltime\n",
    "# Mem Efficiency:  99.5%  1.99 GB of 2.00 GB\n",
    "# Reduce time to 3hrs, CPU can stay at 2, increase memory to 3GB (just to be safe)\n",
    "\n",
    "### Step 3: Bootstrapping PSMC\n",
    "\n",
    "#!/bin/bash -e\n",
    "#SBATCH --job-name=psmc_bootstrap_06Mar2024\n",
    "#SBATCH --account=uoa02613\n",
    "#SBATCH --time=200:00:00\n",
    "#SBATCH --mem=10GB\n",
    "#SBATCH --output=MCN_%j.errout\n",
    "#SBATCH --mail-user=m.nehmens@massey.ac.nz\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=2\n",
    "#SBATCH --profile=task\n",
    "#SBATCH --array=1-12\n",
    "#SBATCH --partition=long\n",
    "\n",
    "# need long partition due to run time, bigmem only runs for 7 days which it automatically selected because of resources. forcing partition still didn't like it, so Dini had to back end fix\n",
    "\n",
    "SAMPLE=$(sed \"${SLURM_ARRAY_TASK_ID}q;d\" /nesi/nobackup/uoa02613/kstuart_projects/At4_MynaStarling/analysis/PSMC/psmc_indv_subset_bs1.txt)\n",
    "\n",
    "cd /nesi/nobackup/uoa02613/kstuart_projects/At4_MynaStarling/analysis/PSMC/\n",
    "module load psmc/0.6.5-gimkl-2018b \n",
    "\n",
    "splitfa ./${SAMPLE}.psmcfa > ./${SAMPLE}-split.psmcfa\n",
    "\n",
    "seq 100 | xargs -i echo psmc -N30 -t5 -r5 -b -p \"4+30*2+4+6+10\" -o ./results/bootstrap/${SAMPLE}.round-{}.psmc ./${SAMPLE}-split.psmcfa | sh\n",
    "\n",
    "\n",
    "##### Plotting PSMC +bootstrapping results #############\n",
    "\n",
    "## need to cat all files to be plotted together\n",
    "cat au03_mai.diploid.psmc au04_mai.diploid.psmc au27_mai.diploid.psmc au05_men.diploid.psmc au06_men.diploid.psmc au15_men.diploid.psmc S_BAD_001M.diploid.psmc S_BAD_007F.diploid.psmc S_BAD_008F.diploid.psmc uk04_nwc.diploid.psmc uk06_nwc.diploid.psmc uk07_nwc.diploid.psmc > combined_diploid.psmc\n",
    "\n",
    "#!/bin/bash -e\n",
    "#SBATCH --job-name=psmc_plot\n",
    "#SBATCH --account=uoa02613\n",
    "#SBATCH --time=1:00:00\n",
    "#SBATCH --mem=4GB\n",
    "#SBATCH --output=MCN_%j.errout\n",
    "#SBATCH --mail-user=mneh623@aucklanduni.ac.nz\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=2\n",
    "#SBATCH --profile=task\n",
    "\n",
    "\n",
    "cd /nesi/nobackup/uoa02613/kstuart_projects/At4_MynaStarling/analysis/PSMC/results/plots\n",
    "\n",
    "module purge\n",
    "module load psmc/0.6.5-gimkl-2018b\n",
    "\n",
    "psmc_plot.pl -u 2.3e-09 -g 2 -M 'au03_mai, au04_mai, au27_mai, au05_men, au06_men, au15_men, S_BAD_001M, S_BAD_007F, S_BAD_008F, uk04_nwc, uk06_nwc, uk07_nwc' combined_psmc_plot /nesi/nobackup/uoa02613/kstuart_projects/At4_MynaStarling/analysis/PSMC/results/combined_diploid.psmc\n",
    "\n",
    "perl epstopdf.pl combined_psmc_plot.eps\n",
    "\n",
    "### concatenate bootstrap rounds (combined bootstrap = cmbdbs)\n",
    "cat au03_mai.round*.psmc > au03_mai_cmbdbs.psmc\n",
    "cat au04_mai.round*.psmc > au04_mai_cmbdbs.psmc\n",
    "cat au27_mai.round*.psmc > au27_mai_cmbdbs.psmc\n",
    "cat au05_men.round*.psmc > au05_men_cmbdbs.psmc\n",
    "cat au06_men.round*.psmc > au06_men_cmbdbs.psmc\n",
    "cat au15_men.round*.psmc > au15_men_cmbdbs.psmc\n",
    "cat S_BAD_001M.round*.psmc > S_BAD_001M_cmbdbs.psmc\n",
    "cat S_BAD_007F.round*.psmc > S_BAD_007F_cmbdbs.psmc\n",
    "cat S_BAD_008F.round*.psmc > S_BAD_008F_cmbdbs.psmc\n",
    "cat uk04_nwc.round*.psmc > uk04_nwc_cmbdbs.psmc\n",
    "cat uk06_nwc.round*.psmc > uk06_nwc_cmbdbs.psmc\n",
    "cat uk07_nwc.round*.psmc > uk07_nwc_cmbdbs.psmc\n",
    "\n",
    "### Plot combined boostrap files \n",
    "psmc_plot.pl -u 2.3e-09 -g 2 ../plots/au03_mai_combBS_plot au03_mai_cmbdbs.psmc\n",
    "psmc_plot.pl -u 2.3e-09 -g 2 ../plots/au04_mai_combBS_plot au04_mai_cmbdbs.psmc\n",
    "psmc_plot.pl -u 2.3e-09 -g 2 ../plots/au27_mai_combBS_plot au27_mai_cmbdbs.psmc\n",
    "psmc_plot.pl -u 2.3e-09 -g 2 ../plots/au05_men_combBS_plot au05_men_cmbdbs.psmc\n",
    "psmc_plot.pl -u 2.3e-09 -g 2 ../plots/au06_men_combBS_plot au06_men_cmbdbs.psmc\n",
    "psmc_plot.pl -u 2.3e-09 -g 2 ../plots/au15_men_combBS_plot au15_men_cmbdbs.psmc\n",
    "psmc_plot.pl -u 2.3e-09 -g 2 ../plots/S_BAD_001M_combBS_plot S_BAD_001M_cmbdbs.psmc\n",
    "psmc_plot.pl -u 2.3e-09 -g 2 ../plots/S_BAD_007F_combBS_plot S_BAD_007F_cmbdbs.psmc\n",
    "psmc_plot.pl -u 2.3e-09 -g 2 ../plots/S_BAD_008F_combBS_plot S_BAD_008F_cmbdbs.psmc\n",
    "psmc_plot.pl -u 2.3e-09 -g 2 ../plots/uk04_nwc_combBS_plot uk04_nwc_cmbdbs.psmc\n",
    "psmc_plot.pl -u 2.3e-09 -g 2 ../plots/uk06_nwc_combBS_plot uk06_nwc_cmbdbs.psmc\n",
    "psmc_plot.pl -u 2.3e-09 -g 2 ../plots/uk07_nwc_combBS_plot uk07_nwc_cmbdbs.psmc\n",
    "\n",
    "### Convert to PDF\n",
    "au03_mai_combBS_plot.eps\n",
    "au04_mai_combBS_plot.eps\n",
    "au27_mai_combBS_plot.eps\n",
    "au05_men_combBS_plot.eps\n",
    "au06_men_combBS_plot.eps\n",
    "au15_men_combBS_plot.eps\n",
    "S_BAD_001M_combBS_plot.eps\n",
    "S_BAD_007F_combBS_plot.eps\n",
    "S_BAD_008F_combBS_plot.eps\n",
    "uk04_nwc_combBS_plot.eps\n",
    "uk06_nwc_combBS_plot.eps\n",
    "uk07_nwc_combBS_plot.eps\n",
    "\n",
    "perl epstopdf.pl au03_mai_combBS_plot.eps\n",
    "perl epstopdf.pl au04_mai_combBS_plot.eps\n",
    "perl epstopdf.pl au27_mai_combBS_plot.eps\n",
    "perl epstopdf.pl au05_men_combBS_plot.eps\n",
    "perl epstopdf.pl au06_men_combBS_plot.eps\n",
    "perl epstopdf.pl au15_men_combBS_plot.eps\n",
    "perl epstopdf.pl S_BAD_001M_combBS_plot.eps\n",
    "perl epstopdf.pl S_BAD_007F_combBS_plot.eps\n",
    "perl epstopdf.pl S_BAD_008F_combBS_plot.eps\n",
    "perl epstopdf.pl uk04_nwc_combBS_plot.eps\n",
    "perl epstopdf.pl uk06_nwc_combBS_plot.eps\n",
    "perl epstopdf.pl uk07_nwc_combBS_plot.eps\n",
    "\n",
    "#### Combine concatenated boostrap rounds with original, just playing here \n",
    "cat  au03_mai.diploid.psmc /nesi/nobackup/uoa02613/kstuart_projects/At4_MynaStarling/analysis/PSMC/results/bootstrap/au03_mai_cmbdbs.psmc > au03_mai_OGBS.psmc\n",
    "cat  au04_mai.diploid.psmc /nesi/nobackup/uoa02613/kstuart_projects/At4_MynaStarling/analysis/PSMC/results/bootstrap/au04_mai_cmbdbs.psmc > au04_mai_OGBS.psmc\n",
    "cat  au27_mai.diploid.psmc /nesi/nobackup/uoa02613/kstuart_projects/At4_MynaStarling/analysis/PSMC/results/bootstrap/au27_mai_cmbdbs.psmc > au27_mai_OGBS.psmc\n",
    "cat  au05_men.diploid.psmc /nesi/nobackup/uoa02613/kstuart_projects/At4_MynaStarling/analysis/PSMC/results/bootstrap/au05_men_cmbdbs.psmc > au05_men_OGBS.psmc\n",
    "cat  au06_men.diploid.psmc /nesi/nobackup/uoa02613/kstuart_projects/At4_MynaStarling/analysis/PSMC/results/bootstrap/au06_men_cmbdbs.psmc > au06_men_OGBS.psmc\n",
    "cat  au15_men.diploid.psmc /nesi/nobackup/uoa02613/kstuart_projects/At4_MynaStarling/analysis/PSMC/results/bootstrap/au15_men_cmbdbs.psmc > au15_men_OGBS.psmc\n",
    "cat  S_BAD_001M.diploid.psmc /nesi/nobackup/uoa02613/kstuart_projects/At4_MynaStarling/analysis/PSMC/results/bootstrap/S_BAD_001M_cmbdbs.psmc > S_BAD_001M_OGBS.psmc\n",
    "cat  S_BAD_007F.diploid.psmc /nesi/nobackup/uoa02613/kstuart_projects/At4_MynaStarling/analysis/PSMC/results/bootstrap/S_BAD_007F_cmbdbs.psmc > S_BAD_007F_OGBS.psmc\n",
    "cat  S_BAD_008F.diploid.psmc /nesi/nobackup/uoa02613/kstuart_projects/At4_MynaStarling/analysis/PSMC/results/bootstrap/S_BAD_008F_cmbdbs.psmc > S_BAD_008F_OGBS.psmc\n",
    "cat  uk04_nwc.diploid.psmc /nesi/nobackup/uoa02613/kstuart_projects/At4_MynaStarling/analysis/PSMC/results/bootstrap/uk04_nwc_cmbdbs.psmc > uk04_nwc_OGBS.psmc\n",
    "cat  uk06_nwc.diploid.psmc /nesi/nobackup/uoa02613/kstuart_projects/At4_MynaStarling/analysis/PSMC/results/bootstrap/uk06_nwc_cmbdbs.psmc > uk06_nwc_OGBS.psmc\n",
    "cat  uk07_nwc.diploid.psmc /nesi/nobackup/uoa02613/kstuart_projects/At4_MynaStarling/analysis/PSMC/results/bootstrap/uk07_nwc_cmbdbs.psmc > uk07_nwc_OGBS.psmc\n",
    "\n",
    "### Plot main PSMC and bootstrap -- dark red line is now main PSMC in the combined plot for all locations\n",
    "psmc_plot.pl -u 2.3e-09 -g 2 au03_mai_OGBS_plot ../au03_mai_OGBS.psmc\n",
    "psmc_plot.pl -u 2.3e-09 -g 2 au04_mai_OGBS_plot ../au04_mai_OGBS.psmc\n",
    "psmc_plot.pl -u 2.3e-09 -g 3 au27_mai_OGBS_plot ../au27_mai_OGBS.psmc\n",
    "psmc_plot.pl -u 2.3e-09 -g 4 au05_men_OGBS_plot ../au05_men_OGBS.psmc\n",
    "psmc_plot.pl -u 2.3e-09 -g 5 au06_men_OGBS_plot ../au06_men_OGBS.psmc\n",
    "psmc_plot.pl -u 2.3e-09 -g 6 au15_men_OGBS_plot ../au15_men_OGBS.psmc\n",
    "psmc_plot.pl -u 2.3e-09 -g 7 S_BAD_001M_OGBS_plot ../S_BAD_001M_OGBS.psmc\n",
    "psmc_plot.pl -u 2.3e-09 -g 8 S_BAD_007F_OGBS_plot ../S_BAD_007F_OGBS.psmc\n",
    "psmc_plot.pl -u 2.3e-09 -g 9 S_BAD_008F_OGBS_plot ../S_BAD_008F_OGBS.psmc\n",
    "psmc_plot.pl -u 2.3e-09 -g 10 uk04_nwc_OGBS_plot ../uk04_nwc_OGBS.psmc\n",
    "psmc_plot.pl -u 2.3e-09 -g 11 uk06_nwc_OGBS_plot ../uk06_nwc_OGBS.psmc\n",
    "psmc_plot.pl -u 2.3e-09 -g 12 uk07_nwc_OGBS_plot ../uk07_nwc_OGBS.psmc\n",
    "\n",
    "### Convert to PDF\n",
    "perl epstopdf.pl au03_mai_OGBS_plot.eps\n",
    "perl epstopdf.pl au04_mai_OGBS_plot.eps\n",
    "perl epstopdf.pl au27_mai_OGBS_plot.eps\n",
    "perl epstopdf.pl au05_men_OGBS_plot.eps\n",
    "perl epstopdf.pl au06_men_OGBS_plot.eps\n",
    "perl epstopdf.pl au15_men_OGBS_plot.eps\n",
    "perl epstopdf.pl S_BAD_001M_OGBS_plot.eps\n",
    "perl epstopdf.pl S_BAD_007F_OGBS_plot.eps\n",
    "perl epstopdf.pl S_BAD_008F_OGBS_plot.eps\n",
    "perl epstopdf.pl uk04_nwc_OGBS_plot.eps\n",
    "perl epstopdf.pl uk06_nwc_OGBS_plot.eps\n",
    "perl epstopdf.pl uk07_nwc_OGBS_plot.eps\n",
    "\n",
    "\n",
    "\n",
    "### Naming key found in /nesi/nobackup/uoa02613/kstuart_projects/At4_MynaStarling/analysis/PSMC/results (any files used prior to this are in directory PSMC, including fastq files)\n",
    "*.sh are my SLURM scripts (did interactive for boostrapping once rounds were created, I wanted to make sure each step did what I wanted it to- didn't take extra time)\n",
    "*.diploid.psmc are the original PSMC files (/results)\n",
    "combined_diploid.psmc combined *.diploid.psmc files (/results)\n",
    "combined_psmc_plot.* the plotted files of the combined_diploid.psmc (./plots)\n",
    "*_cmbdbs.psmc are the 'combined bootstrap' rounds by individaul (./bootstrap)\n",
    "*_combBS_plot.* are the plotted combined bootstrap rounds (./plots)\n",
    "*_OGBS.psmc are the original PSMC (OG) and bootstrap (BS) PSMC files concatenated together (/results)\n",
    "*_OGBS_plot.* are the plotted original PSMC and boostrap (./plots)\n"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
